### ISSUE TITLE
Ray

### ISSUE THREAD

**Issue by YunchaoYang** [YunchaoYang/blogs#65]
1. 什么是Ray: 
Ray是UC Berkeley RISELab新推出的高性能分布式执行框架，具有比Spark更优异的计算性能，而且部署和改造更简单，同时支持机器学习和深度学习的分布式训练，支持主流的深度学习框架（pytorch，tensorflow，keras等）。

2. Ray架构
   - Node: 节点，主要是head和worker， head可以认为是Master，worker是执行任务的单元
   - 每个节点都有自己的本地调度器local scheduler
   - object store：一个内存对象存储，允许Node之间进行通信
   - scheduler：有两个调度器，每个节点都有本地的调度器， 在提交任务时，Local Scheduler会判断是否需要提交给Global 
   
    Scheduler分发给其他worker来执行。
   - GCS：全局状态控制记录了Ray中各种对象的状态信息，可以认为是meta数据，是Ray容错的保证

![image](https://github.com/YunchaoYang/Blogs/assets/6526592/6bd55176-48cb-46c8-9bf6-f81d6238efd4)

Ray适用于任何分布式计算的任务，包括分布式训练。笔者最近是用在大量的时间序列预测模型训练和在线预测上。

3. Ray目前库支持
  - 超参数调优Ray tune， 
  - 梯度下降Ray SGD，
  - 推理服务RaySERVE, 
  - 分布式数据Dataset
  - 以及分布式增强学习RLlib。

4.   基本操作函数
    - Ray 通过`@ray.remote`装饰器使得函数变成可分布式调用的任务。
    - 通过函数名.remote方式进行提交任务，通过ray.get方式来获取任务返回值。
    - remote返回的对象的id 如ObjectRef(7f10737098927148ffffffff0100000001000000)。
    - 需要通过`ray.get`来获取实际的值， 需要注意的是ray.get是阻塞式的调用，不能[ray.get(do_some_work.remote(x)） for x in range(4)]
    - 注意小任务使用情况 需要注意的是ray分布式计算在调度的时候需要发费额外的时间，如调度，进程间通信以及任务状态的更新等等，所以避免过小的任务。可以把小任务进行合并 
    -  ray.put ray.put() 把一个对象放到对象存储上，返回一个object id， 这个id可以在分布式机器上都可以调用，该操作为异步的。通过ray.get()可以是获取。 
    -  ray.wait 如果任务返回多个结果，ray.get()会等所有结果都完成之后才会执行后续的操作。如果多个结果执行的耗时不同，此时短板在于最长的那个任务。
    - 这个时候可以采用ray.wait()方法，ray.wait()返回执行完毕的和未执行完毕的任务结果，执行完成的结果可以继续后续的操作   

5. 集群部署
    - Ray的架构遵循master-slave的模式。Head Node 可以认为是Master，其他的Node为worker。
    - 在集群部署时，Head Node需要首先启动`ray start --head`, 其他机器依次启动worker, 注意需要指定head Node的地址确定关系，`ray start --address 10.8.xx.3:6379`。
    - 关闭服务，需要每一台机器执行 `ray.stop`

6. 任务依赖 
任务之间存在依赖关系，Ray和Spark一样也是通过生成DAG图的方式来确定依赖关系，确定可以并行跑的任务。

有状态Actor以及深度学习支持

**Comment by YunchaoYang** [comment]
How to convert a serial traiing to distribtued multi-worker training: 
1. Use the `ray.train.torch.prepare_model` and `ray.train.torch.prepare_data_loader` utility functions to set up your model and data for distributed training. 
2. This automatically wraps the model with `DistributedDataParallel `and places it on the right device, and adds `DistributedSampler `to the DataLoaders.
